### PR TITLE
Override `dataclass_transform` behavior for `RootModel`

### PR DESCRIPTION
**Pull request by Viicos**
## Change Summary

Fixes https://github.com/pydantic/pydantic/discussions/7418.

Seems to work as expected on mypy (you can experiment [here](https://mypy-play.net/?mypy=1.2.0&python=3.11&gist=cc7e52cead578c3844cfa4a92c92b2d0)).

For pyright, see the comment below.

There's some mypy tests for root models, but they seem to only run on mypy 1.0.1 and with the plugin.

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @hramezani

### CODE REVIEW

**Comment by Viicos** [comment]
I don't know if this is still required in the mypy plugin if this PR gets merged?

https://github.com/pydantic/pydantic/blob/24272b4f856829dc10660d3f9e1f4607246afa84/pydantic/mypy.py#L860-L862

File: pydantic/root_model.py [hunk 1]
```diff
@@ -25,6 +27,7 @@
 RootModelRootType = typing.TypeVar('RootModelRootType')
 
 
+@dataclass_transform(kw_only_default=False, field_specifiers=(PydanticModelField,))
 class RootModel(BaseModel, typing.Generic[RootModelRootType]):
     """Usage docs: https://docs.pydantic.dev/2.6/concepts/models/#rootmodel-and-custom-root-types
```
Comment (Viicos) [inline pydantic/root_model.py:32]: Although this works with `mypy` (e.g. creating subclass of `RootModel` will allow positional argument for `root`), pyright does not support it (see: https://github.com/microsoft/pyright/issues/6489#issuecomment-1817642695).

One workaround would be to have a metaclass for `RootModel`, subclassing `ModelMetaclass`, and we could set `dataclass_transform.kw_only_default=False` on this metaclass.

A fix to support having `dataclass_transform` applied directly on `RootModel` is also available, so this is up to you (see issue on pyright for more context).

**Comment by Viicos** [comment]
Please review

**Comment by dmontagu** [comment]
My initial reaction to this is that:

* If it is possible to do something with a sub-metaclass of `ModelMetaclass`, I would make a `RootModelMetaclass` in an `if TYPE_CHECKING:` block, and in the `else:`, just set it to be exactly `ModelMetaclass`. Something like:

```diff
diff --git a/pydantic/root_model.py b/pydantic/root_model.py
diff --git a/pydantic/root_model.py b/pydantic/root_model.py
index 97381233..a0c0bf00 100644
--- a/pydantic/root_model.py
+++ b/pydantic/root_model.py
@@ -8,15 +8,23 @@ from copy import copy, deepcopy
 from pydantic_core import PydanticUndefined
 
 from . import PydanticUserError
-from ._internal import _repr
+from ._internal import _model_construction, _repr
 from .main import BaseModel, _object_setattr
 
 if typing.TYPE_CHECKING:
-    from typing import Any
+    from typing import Any, dataclass_transform
 
     from typing_extensions import Literal
 
+    from .fields import Field as PydanticModelField
+
+    @dataclass_transform(kw_only_default=False, field_specifiers=(PydanticModelField,))
+    class _RootModelMetaclass(_model_construction.ModelMetaclass, typing.Generic[typing.Any]):
+        ...
+
     Model = typing.TypeVar('Model', bound='BaseModel')
+else:
+    _RootModelMetaclass = _model_construction.ModelMetaclass
 
 
 __all__ = ('RootModel',)
@@ -25,7 +33,7 @@ __all__ = ('RootModel',)
 RootModelRootType = typing.TypeVar('RootModelRootType')
 
 
-class RootModel(BaseModel, typing.Generic[RootModelRootType]):
+class RootModel(BaseModel, typing.Generic[RootModelRootType], metaclass=_RootModelMetaclass):
     """Usage docs: https://docs.pydantic.dev/2.6/concepts/models/#rootmodel-and-custom-root-types
 
     A Pydantic `BaseModel` for the root object of the model.
```
I think it's best to do it in the `if TYPE_CHECKING:` block to keep things as performant as possible and minimize the risk that there is any possible impact on runtime behavior of existing code.

* If this causes issues with mypy or pyright (I haven't checked if it does, but I did check that it works without issue at runtime), I think we should maybe go back and ask for the change to pyright to make the change in pydantic unnecessary.

**Comment by dmontagu** [comment]
Regarding the point:

> I don't know if this is still required in the mypy plugin if this PR gets merged?

I would suggest we remove that logic branch as part of this PR if we can and mypy still handles root models correctly (due to the change to the dataclass_transform). I strongly suspect we have code in one of the existing mypy tests that checks that you don't get an error with `RootModel` if you use a positional argument to the `__init__`, but it would be good to confirm this prior to merging changes to the plugin code (assuming those changes don't cause any tests to fail).

File: tests/mypy/outputs/1.0.1/mypy-plugin_ini/root_models.py [hunk 2]
```diff
@@ -16,8 +16,10 @@ class Pets3(RootModel):
 
 
 pets1 = Pets1(['dog', 'cat'])
+# MYPY: error: Too many positional arguments for "Pets1"  [misc]
 pets2 = Pets2(['dog', 'cat'])
 pets3 = Pets3(['dog', 'cat'])
+# MYPY: error: Too many positional arguments for "Pets3"  [misc]
```
Comment (Viicos) [inline tests/mypy/outputs/1.0.1/mypy-plugin_ini/root_models.py:22]: mypy support for dataclass_transform isn't really good iirc, hence the errors

**Comment by Viicos** [comment]
Please review

File: pydantic/mypy.py [hunk 3]
```diff
@@ -857,9 +855,6 @@ def add_initializer(
                 use_alias=use_alias,
                 is_settings=is_settings,
             )
-            if is_root_model:
-                # convert root argument to positional argument
-                args[0].kind = ARG_POS if args[0].kind == ARG_NAMED else ARG_OPT
```
Comment (sydney-runkle) [inline pydantic/mypy.py:862]: @Viicos,

If we add this back, do we get no errors for `mypy 1.0.1` as well as all of the other `mypy` versions currently tested?

If so, I'd prefer to do that + add a note about the special case for 1.0.1.

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
@Viicos,

Looks great otherwise, just wondering if we can avoid reintroducing `mypy` errors in `1.0.1` if possible.

File: pydantic/mypy.py (hunk 3 above)
Comment (sydney-runkle) [reply pydantic/mypy.py:862]: Or if necessary, wrap this check in a block that checks the `mypy` version

File: pydantic/mypy.py [hunk 4]
```diff
@@ -860,8 +860,10 @@ def add_initializer(
                 use_alias=use_alias,
                 is_settings=is_settings,
             )
-            if is_root_model:
+
+            if is_root_model and MYPY_VERSION_TUPLE <= (1, 0, 1):
                 # convert root argument to positional argument
+                # This is needed because mypy support for `dataclass_transform` isn't complete on 1.0.1
```
Comment (sydney-runkle) [inline pydantic/mypy.py:866]: Great, just what I was hoping for. Thanks 👍

File: tests/mypy/outputs/1.0.1/mypy-plugin_ini/root_models.py (hunk 2 above)
Comment (sydney-runkle) [reply tests/mypy/outputs/1.0.1/mypy-plugin_ini/root_models.py:22]: @Viicos, can't we remove these now that you added back support for 1.0.1?

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
Looks great otherwise. Please see my question on the 1.0.1 tests. Otherwise, ready to approve and merge 👍

**Review by sydney-runkle — APPROVED** [review]
Great! Approved.

**Pull request merged**